src/remote/build_metadata_store.go

```diff
@@ -0,0 +1,125 @@
+package remote
+
+import (
+	"encoding/gob"
+	"fmt"
+	"github.com/thought-machine/please/src/core"
+	"github.com/thought-machine/please/src/fs"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+const pleaseCacheDirName = "please"
+const storeDirectoryName = "build-metadata-store"
+
+type buildMetadataStore interface {
+	storeMetadata(key string, md *core.BuildMetadata) error
+	retrieveMetadata(key string) (*core.BuildMetadata, error)
+}
+
+type directoryMetadataStore struct {
+	directory     string
+	cacheDuration time.Duration
+}
+
+func newDirMDStore(cacheDuration time.Duration) *directoryMetadataStore {
+	userCacheDir, err := os.UserCacheDir()
+	if err != nil {
+		log.Fatalf("failed to find user cache dir for metadata store: %v", err)
+	}
+	dir := filepath.Join(userCacheDir, pleaseCacheDirName, storeDirectoryName)
+
+	if err := os.MkdirAll(dir, fs.DirPermissions); err != nil {
+		log.Fatalf("failed to create metadata store directory: %v", err)
+	}
+	store := &directoryMetadataStore{
+		directory:     dir,
+		cacheDuration: cacheDuration,
+	}
+
+	go store.clean()
+	return store
+}
+
+// clean will delete any cached metadata that has expired
+func (d *directoryMetadataStore) clean() {
+	_ = fs.Walk(d.directory, func(name string, isDir bool) error {
+		if isDir {
+			return nil
+		}
+
+		if md, err := loadMetadata(name); err == nil && d.hasExpired(md) {
```

Comment:
you might need to make sure this isn't trying to read a file you're still in the process of writing? or do we just think the `err == nil` will solve that?
I think it's atomic on the read side...

Reply:
This is all best effort. I assume it will error and just carry on happily nuking as much as it can.

Reply:
I *think* this is ok because if it's still  being written it will error  on read and won't get inside this loop...